### ISSUE TITLE
🔥 [🐛] ignoreUndefinedProperties not respected for arrays of objects

### ISSUE THREAD

**Issue by pehagg** [invertase/react-native-firebase#5437]
## Issue

According to what I'm seeing, the `ignoreUndefinedProperties` setting is not respected for objects in arrays, i.e. if you have an object in an array that has one or more undefined values, then you'll get "Unsupported field value: undefined", even if you've set `ignoreUndefinedProperties` to `true`.

I tracked down the issue and it could be fixed by adding `ignoreUndefined` to `buildNativeArray` and passing it along to `generateNativeData`, as well as adding the argument to the `buildNativeArray` call in `generateNativeData` at line 108 (or thereabouts).

Here lies a hidden question, should the arrays be checked like this or not? I'm not familiar with how the JS SDK behaves and nor do I have the time or the interest too look into it right now.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
# N/A
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
 OUTPUT GOES HERE
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [x] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `12.1.0`
- **`Firebase` module(s) you're using that has the issue:**
  - `Firestore`
- **Are you using `TypeScript`?**
  - `N/A`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
interesting! I can't imagine why we'd have a toggle to ignore them and still see that message ever, I'll be surprised if firebase-js-sdk works that way. Marked for review but I'm guessing you are correct and I appreciate the solution sketch

**Comment by pehagg** [comment]
> [...] I can't imagine why we'd have a toggle to ignore them and still see that message ever, I'll be surprised if firebase-js-sdk works that way. [...]

That's my assumption as well, but believe me, I've seen some really weird SDKs during my 20 years as a developer.

**Comment by mikehardy** [comment]
> > [...] I can't imagine why we'd have a toggle to ignore them and still see that message ever, I'll be surprised if firebase-js-sdk works that way. [...]
> 
> That's my assumption as well, but believe me, I've seen some really weird SDKs during my 20 years as a developer.

Preach :pray: :laughing:

**Comment by RayVR** [comment]
this is requiring a decent amount of work to deal with. Is this planned for a BF release any time soon or should I downgrade to the last version that did not have this issue?

**Comment by mikehardy** [comment]
Hey @RayVR or @pehagg - I'm investigating this and want to add a test case to make sure I get it right. I'm pretty sure I understand it well but just to make sure could one of you provide a sample JSON object that unexpectedly fails when you have ignoreUndefinedProperties true and send it in?

**Comment by TMcLoone** [comment]
@mikehardy Here's an example of something that fails for me: 

`[{name: 'Tim',  location: { state: undefined, country: 'United Kingdom'} }] `

I haven't tested it to see if it fails when it's not nested but by the sounds of the others it does.

**Comment by mikehardy** [comment]
About as simple as I expected ;-) but happy to have a known-bad example, thank you @TMcLoone !

**Comment by pehagg** [comment]
Sorry for not replying earlier, haven't been at the keyboard since Thursday noon. Can confirm that @TMcLoone 's example is in-line with our code that triggers the issue.

**Comment by RayVR** [comment]
@mikehardy Glad to hear there's a new testcase for this. What is the bug fix release cycle like for this project? I'm unfortunately running into this issue frequently in unexpected places.

**Comment by mikehardy** [comment]
Sorry there's no published fix yet. As with all open source the cycle is either someone proposes a fix from the community and I can merge and release in the same day plus there's an auto generated set of patch package patches, or it waits on me or another maintainer with no ETA given. The feature was added recently as well, it may be trivial to simply go in and reverse it or use the older version temporarily

Without giving an ETA as those simply lead to disappointment in open source I can say it's personally a high priority and having a reproduction is typically the biggest hurdle. We already have that

**Comment by mikehardy** [comment]
Your patience has been greatly appreciated! I'm slowly catching up over here in firebase-land and I think I've got this one. Most of the PR I just posted is testing support (thanks again for the sample, obvious though it is, it was useful to be sure)

I'd appreciate if someone could check the PR and maybe integrate the patch-package set to verify? It's the first patch post-release so it's super clean - 

- https://github.com/invertase/react-native-firebase/suites/3264083106/artifacts/75965779
- from https://github.com/invertase/react-native-firebase/pull/5527/checks?check_run_id=3093917547

**Comment by mikehardy** [comment]
This seemed important and is a real problem but I haven't heard back from anyone? In the absence of confirmation I will assume my fix is good and will merge and release it later today

**Comment by RayVR** [comment]
Mike, 
Thank you. This is important as I have a half-baked solution to avoid crashes in production but a real solution will restore functionality.

Thank you!

**Comment by pehagg** [comment]
@mikehardy , sorry for not replying earlier. I've been away from my computer for the last three weeks. I'll have a look at this later today.

**Comment by mikehardy** [comment]
No problem, hopefully I nailed the issue, it's released already but we can always do a follow up PR if needed

**Comment by pehagg** [comment]
I'm planning to update our libraries next, will verify the fix within the next 24 hours or so. Got a little bit delayed by other stuff after returning from the vacays.

**Comment by pehagg** [comment]
Can confirm that the fix is working when using version 12.3.0.
